High level actions humanoid

xavierpuigf opened the pull request:
## Motivation and Context

This PR allows to control humanoids using high level skills. We currently only provide an oracle navigation skill
for navigation and Noop skills for pick and place actions.  It works both with PDDL policies as well as with trained policies.

The PR updates the humanoid controller to enable good navigation. In particular it corrects the rotations that happen to the humanoid because of gait to enable a smooth path.

## How Has This Been Tested

Works using a PDDL-based policy. 
Works using Trained Policy. Tested checking video.

## Types of changes

- **[Docs change]** Addition to changes in documentation.
- **[New feature]** (non-breaking change which adds functionality)
- **[Development]** A pull request that add new features to the [habitat-lab](https://github.com/habitat-lab) task and environment codebase.

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

ASzot commented on habitat-baselines/habitat_baselines/config/rearrange/rl_hierarchical_oracle_nav_human.yaml:
```diff
@@ -0,0 +1,58 @@
+# @package _global_
```
I don't like creating another one of these configs. Is there any way you can unify this with the existing `rl_hierarchical` config and have argument overrides or compose with another config to configure the human? At the very least, you should derive from the existing `rl_hierarchical_oracle_nav` config.

ASzot commented on habitat-lab/habitat/config/habitat/task/rearrange/rearrange_easy_humanoid.yaml:
```diff
@@ -0,0 +1,48 @@
+# @package habitat.task
```
Same comment as above. I think we should not create another config for another agent. Can you compose configs or use argument overrides to achieve this?

xavierpuigf replied:
I started working on this, but I don't know how to remove the arm_action for the humanoid, that appears here: https://github.com/facebookresearch/habitat-lab/blob/1622862c53017376f6f7760c549e08b7efd49870/habitat-lab/habitat/config/habitat/task/rearrange/rearrange_easy.yaml#L45.

Even if I do override with the actions, arm_action appears and fails with the humanoid. Any idea to address this?

ASzot commented on habitat-lab/habitat/core/env.py:
```diff
@@ -128,6 +128,28 @@ def __init__(
         )
         self.action_space = self._task.action_space
+
+        # Sharing controller between actions
```
This is loading the human controller for all environments? This should definitely go elsewhere. Why not in the action controller?

xavierpuigf replied:
Where is the action controller?

xavierpuigf replied:
Thanks for the feedback! Addressed in latest commit.

vincentpierre commented on habitat-baselines/habitat_baselines/articulated_agent_controllers/humanoid_rearrange_controller.py:
```diff
@@ -161,9 +187,10 @@ def get_walk_pose(
+                new_angle = new_angle * np.pi / 180
 
             forward_V = mn.Vector3(np.sin(new_angle), 0, np.cos(new_angle))
+            new_angle = new_angle * 180 / np.pi
```
Why are we setting a value for new_angle here? It does not look like it is used anywhere later in the code.

xavierpuigf replied:
Addressed in last commit

vincentpierre commented on habitat-baselines/habitat_baselines/articulated_agent_controllers/humanoid_rearrange_controller.py:
```diff
@@ -225,11 +252,12 @@ def get_walk_pose(
-        obj_transform.translation += (
-            forward_V * dist_diff * distance_multiplier
-        )
+        # This is the rotation and translation explained by the current pose
```
is explained the correct word ?

xavierpuigf replied:
Addressed in last commit

vincentpierre commented on habitat-baselines/habitat_baselines/config/default_structured_configs.py:
```diff
@@ -243,6 +243,7 @@ class HrlDefinedSkillConfig(HabitatBaselinesBaseConfig):
     # If true, this willapply the post-conditions of the skill after it
     # terminates.
     apply_postconds: bool = False
+    ignore_grip: bool = False
```
what does this do ?

xavierpuigf replied:
Added a comment in default_structured_config.py

vincentpierre commented on habitat-baselines/habitat_baselines/config/habitat_baselines/rl/policy/hierarchical_policy/defined_skills/oracle_skills_nogrip.yaml:
```diff
@@ -0,0 +1,67 @@
+# Oracle skills that will teleport to the skill post-condition. When automatically setting predicates you may want to run the simulation in kinematic mode:
```
Can you add an explanation to what ignore_grip changed here ?

xavierpuigf replied:
Added a comment in default_structured_config.py

ASzot commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -107,47 +129,97 @@ def step(self, *args, is_last_action, **kwargs):
-
-        angle_to_target = get_angle(robot_forward, rel_targ)
-        angle_to_obj = get_angle(robot_forward, rel_pos)
+        if self.motion_type != "base_velocity":
```
The `step` function is now getting a too long and complicated. Can you break this up into sub-methods to make it clear what the individual components are doing?

xavierpuigf replied:
I did some changes that reduce it. Let me know if that is still too long.

ASzot reviewed (changes requested):
See above comments.

vincentpierre commented on habitat-lab/habitat/core/env.py:
```diff
@@ -128,6 +128,28 @@ def __init__(
+        # This should go somewhere else, but it is unclear where
+        # we want to share a controller across actions since the controller
+        # is stateful
+        from habitat_baselines.articulated_agent_controllers import (
```
Absolutely not. habitat-baselines cannot be a dependency of habitat-lab.

xavierpuigf replied:
Totally agreed. I need actions to share the humanoid controller and I was not sure the best place to set them. I am looking at whether this can be added in actions controller. Otherwise, this is a good argument for moving the humanoid controllers into Habitat-Lab, as suggested in the previous PR.

xavierpuigf replied:
Addressed in latest commit. There is a separate PR for that here https://github.com/facebookresearch/habitat-lab/pull/1202

vincentpierre commented on habitat-lab/habitat/core/env.py:
```diff
@@ -128,6 +128,28 @@ def __init__(
+            HumanoidRearrangeController,
+        )
+
+        walk_pose_path = (
```
put this code in a helper method to isolate it better.

xavierpuigf replied:
Addressed in latest commit

vincentpierre commented on habitat-lab/habitat/tasks/rearrange/actions/actions.py:
```diff
@@ -558,16 +558,19 @@ def action_space(self):
         r"""
         Updates the joint rotations and root transformation of the humanoid.
-        :param human_joint_trans: Array of size (num_joints*4)+16. The last 16
+        :param human_joints_trans: Array of size (num_joints*4)+16. The last 16
```
according to line 571, this is not the correct parameter name.

xavierpuigf replied:
Sorry, I don't understand this comment. Do you mean I should remove this parameter, and only document `is_last_action`?

vincentpierre replied:
I mean that the correct parameter name is `self._action_arg_prefix + "human_joints_trans"`

eundersander commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -107,47 +129,97 @@ def step(self, *args, is_last_action, **kwargs):
-        angle_to_target = get_angle(robot_forward, rel_targ)
-        angle_to_obj = get_angle(robot_forward, rel_pos)
+        if self.motion_type != "base_velocity":
+            # The humanoid has a root rotation and translation when moving
```
This code is not very readable because it's not clear what type of `motion_type` you're handling in this branch. You are expecting `motion_type` to be `human_joints`, is that right? Can you add an assert or otherwise do some error-handling?

xavierpuigf replied:
Resolved in last commit

eundersander commented on habitat-lab/habitat/tasks/rearrange/actions/oracle_nav_action.py:
```diff
@@ -3,30 +3,41 @@
+        if self.motion_type == "base_velocity":
+            BaseVelAction.__init__(self, *args, **kwargs)
+        else:
+            self.humanoid_controller = None
```
See my other comment about making this more readable for the `human_joints` case.

xavierpuigf replied:
Thanks for the review! Provided better handling in latest commit.

eundersander commented on habitat-lab/habitat/core/env.py:
```diff
@@ -128,6 +128,28 @@ def __init__(
+        )
+
+        walk_pose_path = (
+            "data/humanoids/humanoid_data/walking_motion_processed.pkl"
```
This looks strange to me. We rarely (never?) hard-code data paths in python code.

eundersander replied:
Also, is this file needed for this code to run? How do users get it?

xavierpuigf replied:
Right! I will change it so that it is part of the humanoid config. The plan is to release this data with Habitat3.0 but there are some steps we need to do in order to release this first.

xavierpuigf replied:
Addressed in the latest commit

eundersander commented on habitat-lab/habitat/core/env.py:
```diff
@@ -128,6 +128,28 @@ def __init__(
+            "data/humanoids/humanoid_data/walking_motion_processed.pkl"
+        )
+        humanoid_controller = HumanoidRearrangeController(walk_pose_path)
+        if (
```
At a glance, `core/env.py` looks to be task-agnostic code. It seems hacky to have this task-specific logic in here.

xavierpuigf replied:
Addressed in the latest commit

eundersander reviewed (approved):
I left some comments.

xavierpuigf commented:
I addressed all the comments, let me know what you think.

I simplified the config structure, but I didn't find a way to delete an action that had been defined in a config before. This is why I created a `rearrange_easy_base.yaml` that is exactly like the previous [rearrange_easy.yaml](https://github.com/facebookresearch/habitat-lab/blob/main/habitat-lab/habitat/config/habitat/task/rearrange/rearrange_easy.yaml) but without the arm_action, and changed rearrange_easy to derive from it and add an arm_action. I am not sure if there is a simpler way to do this.

vincentpierre reviewed (approved):
I left some more comments on #1202

xavierpuigf commented:
> See above comments.

Thanks for the feedback! I think I addressed all the issues. Let me know.

ASzot reviewed (approved):
LGTM.
